{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bff91b09",
   "metadata": {},
   "source": [
    "# Exercises Week 11-12: Graded exercise session (part B)\n",
    "\n",
    "**Course**: [Topics in life sciences engineering](https://moodle.epfl.ch/enrol/index.php?id=17061) (BIO-411)\n",
    "\n",
    "**Professors**:  _GÃ¶nczy Pierre_, _Naef Felix_, _McCabe Brian Donal_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "867ce50e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.decomposition import PCA\n",
    "import plotly.express as px\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import scale\n",
    "import gseapy as gp\n",
    "\n",
    "from scipy.integrate import odeint\n",
    "import matplotlib.pyplot as plt\n",
    "from ipywidgets import interact, fixed\n",
    "from scipy.stats import beta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e04f778e",
   "metadata": {},
   "source": [
    "**Provide answers in the form of code, figures (if relevant) and short descriptions (in markdown cells) in those notebooks. Submit your notebook to Moodle, please make sure to execute every cell.**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4dcb272",
   "metadata": {},
   "source": [
    "### Exercise 1: Simulation of circadian gene regulation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecaa0b5a",
   "metadata": {},
   "source": [
    "Similarly to Week 9 exercises, we first describe a system in which a (nuclear) pre-mRNA is transcribed and spliced to produce an mRNA. We can write a 2D ODE (ignoring the degradation of the pre-mRNA):\n",
    "\n",
    "\\begin{eqnarray*}\n",
    "&&\\frac{dP}{dt} = s(t) - \\rho P \\\\\n",
    "&&\\frac{dM}{dt} = \\rho P - k(t) M \\\\\n",
    "\\end{eqnarray*}\n",
    "\n",
    "Here, $P$ and $M$ denote, respectively, the concentrations of pre-mRNA and mRNA. We will consider time-dependent (circadian) transcription and degradation rates. The frequency is $\\omega=\\frac{2\\pi}{T}$ with $T=24h$. $s(t)$ is the transcription rate, which is now taken either as constant $s(t)=s_{0}$ or as a rhythmic function of time:  \n",
    "\n",
    "\\begin{equation*}\n",
    "s(t) = s_0 + (1 + \\epsilon_s \\cos(\\omega t)) \n",
    "\\end{equation*}  \n",
    "with relative amplitude $0\\leq \\epsilon_s\\leq 1$. Note that here the peak of $s(t)$ is at $t=0$. \n",
    "Similarly, we will consider $k(t)$ either as constant $k(t)= k_0$ or a rhythmic function of time:  \n",
    "\\begin{equation*}\n",
    "k(t)=k_0(1 + \\epsilon_k \\cos(\\omega (t-t_k))\n",
    "\\end{equation*}  \n",
    "with $0\\leq\\epsilon_k\\leq 1$, and the maximum degradation rate at $t=t_k$.\n",
    "\n",
    "$\\rho$ is the splicing rate of pre-mRNA $P$ to mRNA $M$, and is also taken as a constant."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebecf369",
   "metadata": {},
   "source": [
    "### Question 1\n",
    "1. Adapt the code of week 9 exercises to simulate this new system. Modify the interactive widget such that you can vary the new parameters and plot $P(t)$ and $M(t)$.  \n",
    "2. Keep the degradation rate constant ($\\epsilon_k$=0) and vary $k_0$ and $\\epsilon_s$. Discuss on how 1) the phase delay between $M(t)$ and $P(t)$ and 2) the relative amplitude of $M(t)$ varies in function of $k_0$. What is the maximal relative amplitude? Make sure you span a relevant range for $k_0^{-1}$ ( *i.e.* from 10 minutes to several hours).\n",
    "3. Vary $\\epsilon_k$ to include rhythmic degradation. Discuss the different effects, in particular, find situations that extend beyond what is possible with a constant half-life. See also [Wang et al., 2018](https://www.pnas.org/doi/10.1073/pnas.1715225115)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94467283",
   "metadata": {},
   "source": [
    "### Exercise 2: Circadian (post-)transcriptional regulation of gene expression in mouse liver"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d4c6d92",
   "metadata": {},
   "source": [
    "We will now analyse published RNA-seq data from mouse liver under an *ad libitum* feeding regimen and in presence of a 12h-12h light-dark cycle ([Atger et al., 2015](https://www.pnas.org/doi/abs/10.1073/pnas.1515308112)). The samples were harvested every 2 hours in four replicates, RNA was extracted and sequenced. Similarly to week 9 exercises on RNA-seq data, gene expression was quantified at the intron and exon levels. Note that data are $log_2$ transformed and normalized (RPKM)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcf58338",
   "metadata": {},
   "source": [
    "### Question 1\n",
    "1. Adapt code from week 9 exercises to perform PCA using the 14 following genes (core clock genes and selected clock output genes): ``['Arntl', 'Npas2', 'Clock', 'Cry1', 'Cry2', 'Per1', 'Per2', 'Per3', 'Nr1d1', 'Nr1d2', 'Rora', 'Rorc', 'Tef', 'Dbp']``. First start with only the exons, then only the introns, and eventually both. Describe your observations.\n",
    "2. Using the *return_amp_phase_pv* function with the $log_2$ data, assess rhythmicity genome-wide (use the p-value to make gene selections). For the selected genes, provide histograms of the peak times, and amplitudes ($log_2$ peak-to-trough). Do this separately for the exons and introns. Do you see what's commonly referred to as the morning and evening waves of genes expression?\n",
    "3. Which genes are overall the most robustly rhythmic? Is there a theme?\n",
    "3. Adapting the enrichR code from Week 9 exercises, find biological function or transcriptional programs regulated rhythmically in the mouse liver. Do this separately for the night and the day. Comment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d725b9ef",
   "metadata": {},
   "source": [
    "#### Load the RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0a51a74b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>condition</th>\n",
       "      <th colspan=\"21\" halign=\"left\">WT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feeding</th>\n",
       "      <th colspan=\"21\" halign=\"left\">AL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th colspan=\"10\" halign=\"left\">Intron</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">RFP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>time</th>\n",
       "      <th>00</th>\n",
       "      <th>02</th>\n",
       "      <th>04</th>\n",
       "      <th>06</th>\n",
       "      <th>08</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>14</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>04</th>\n",
       "      <th>06</th>\n",
       "      <th>08</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>14</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>20</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>replicate</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>...</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gnai3</th>\n",
       "      <td>-1.564402</td>\n",
       "      <td>-1.536626</td>\n",
       "      <td>-1.931707</td>\n",
       "      <td>-1.458354</td>\n",
       "      <td>-1.575970</td>\n",
       "      <td>-1.654038</td>\n",
       "      <td>-1.550707</td>\n",
       "      <td>-1.564596</td>\n",
       "      <td>-1.211372</td>\n",
       "      <td>-1.654187</td>\n",
       "      <td>...</td>\n",
       "      <td>4.739938</td>\n",
       "      <td>4.648414</td>\n",
       "      <td>4.576937</td>\n",
       "      <td>4.891934</td>\n",
       "      <td>5.284247</td>\n",
       "      <td>5.502545</td>\n",
       "      <td>5.115915</td>\n",
       "      <td>5.402904</td>\n",
       "      <td>5.119890</td>\n",
       "      <td>4.873039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cdc45</th>\n",
       "      <td>-3.838233</td>\n",
       "      <td>-4.132913</td>\n",
       "      <td>-4.023130</td>\n",
       "      <td>-3.782926</td>\n",
       "      <td>-3.971205</td>\n",
       "      <td>-3.898270</td>\n",
       "      <td>-3.837377</td>\n",
       "      <td>-3.907227</td>\n",
       "      <td>-4.002231</td>\n",
       "      <td>-3.837719</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.783427</td>\n",
       "      <td>-1.045727</td>\n",
       "      <td>-1.868940</td>\n",
       "      <td>-1.616297</td>\n",
       "      <td>1.552717</td>\n",
       "      <td>0.336330</td>\n",
       "      <td>-1.370050</td>\n",
       "      <td>-0.308212</td>\n",
       "      <td>-1.218384</td>\n",
       "      <td>-0.541827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Apoh</th>\n",
       "      <td>-1.075680</td>\n",
       "      <td>-1.540220</td>\n",
       "      <td>-1.569142</td>\n",
       "      <td>-1.652602</td>\n",
       "      <td>-1.786800</td>\n",
       "      <td>-1.539642</td>\n",
       "      <td>-1.434320</td>\n",
       "      <td>-1.655522</td>\n",
       "      <td>-1.499171</td>\n",
       "      <td>-1.411709</td>\n",
       "      <td>...</td>\n",
       "      <td>10.657651</td>\n",
       "      <td>10.460193</td>\n",
       "      <td>10.623698</td>\n",
       "      <td>10.381588</td>\n",
       "      <td>10.146668</td>\n",
       "      <td>10.232355</td>\n",
       "      <td>10.631922</td>\n",
       "      <td>10.468625</td>\n",
       "      <td>10.704931</td>\n",
       "      <td>10.591210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Narf</th>\n",
       "      <td>-1.198793</td>\n",
       "      <td>-0.937221</td>\n",
       "      <td>-0.921084</td>\n",
       "      <td>-0.331866</td>\n",
       "      <td>-0.123671</td>\n",
       "      <td>-0.186613</td>\n",
       "      <td>-0.632810</td>\n",
       "      <td>-0.963087</td>\n",
       "      <td>-0.999763</td>\n",
       "      <td>-1.320254</td>\n",
       "      <td>...</td>\n",
       "      <td>1.788997</td>\n",
       "      <td>2.660656</td>\n",
       "      <td>2.890531</td>\n",
       "      <td>3.379632</td>\n",
       "      <td>2.924193</td>\n",
       "      <td>2.472548</td>\n",
       "      <td>2.686118</td>\n",
       "      <td>2.390108</td>\n",
       "      <td>1.941836</td>\n",
       "      <td>2.128767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cav2</th>\n",
       "      <td>-3.046334</td>\n",
       "      <td>-3.741371</td>\n",
       "      <td>-3.364087</td>\n",
       "      <td>-3.063080</td>\n",
       "      <td>-3.374724</td>\n",
       "      <td>-2.992739</td>\n",
       "      <td>-2.985271</td>\n",
       "      <td>-2.991166</td>\n",
       "      <td>-2.954387</td>\n",
       "      <td>-3.052015</td>\n",
       "      <td>...</td>\n",
       "      <td>0.759856</td>\n",
       "      <td>0.403964</td>\n",
       "      <td>0.581452</td>\n",
       "      <td>0.716145</td>\n",
       "      <td>0.742948</td>\n",
       "      <td>0.385111</td>\n",
       "      <td>0.282426</td>\n",
       "      <td>0.624375</td>\n",
       "      <td>0.806800</td>\n",
       "      <td>0.524357</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 144 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "condition          WT                                                    \\\n",
       "feeding            AL                                                     \n",
       "feature        Intron                                                     \n",
       "time               00        02        04        06        08        10   \n",
       "replicate           A         A         A         A         A         A   \n",
       "Gene_Symbol                                                               \n",
       "Gnai3       -1.564402 -1.536626 -1.931707 -1.458354 -1.575970 -1.654038   \n",
       "Cdc45       -3.838233 -4.132913 -4.023130 -3.782926 -3.971205 -3.898270   \n",
       "Apoh        -1.075680 -1.540220 -1.569142 -1.652602 -1.786800 -1.539642   \n",
       "Narf        -1.198793 -0.937221 -0.921084 -0.331866 -0.123671 -0.186613   \n",
       "Cav2        -3.046334 -3.741371 -3.364087 -3.063080 -3.374724 -2.992739   \n",
       "\n",
       "condition                                            ...             \\\n",
       "feeding                                              ...              \n",
       "feature                                              ...        RFP   \n",
       "time               12        14        16        18  ...         04   \n",
       "replicate           A         A         A         A  ...          D   \n",
       "Gene_Symbol                                          ...              \n",
       "Gnai3       -1.550707 -1.564596 -1.211372 -1.654187  ...   4.739938   \n",
       "Cdc45       -3.837377 -3.907227 -4.002231 -3.837719  ...  -1.783427   \n",
       "Apoh        -1.434320 -1.655522 -1.499171 -1.411709  ...  10.657651   \n",
       "Narf        -0.632810 -0.963087 -0.999763 -1.320254  ...   1.788997   \n",
       "Cav2        -2.985271 -2.991166 -2.954387 -3.052015  ...   0.759856   \n",
       "\n",
       "condition                                                                      \\\n",
       "feeding                                                                         \n",
       "feature                                                                         \n",
       "time                06         08         10         12         14         16   \n",
       "replicate            D          D          D          D          D          D   \n",
       "Gene_Symbol                                                                     \n",
       "Gnai3         4.648414   4.576937   4.891934   5.284247   5.502545   5.115915   \n",
       "Cdc45        -1.045727  -1.868940  -1.616297   1.552717   0.336330  -1.370050   \n",
       "Apoh         10.460193  10.623698  10.381588  10.146668  10.232355  10.631922   \n",
       "Narf          2.660656   2.890531   3.379632   2.924193   2.472548   2.686118   \n",
       "Cav2          0.403964   0.581452   0.716145   0.742948   0.385111   0.282426   \n",
       "\n",
       "condition                                     \n",
       "feeding                                       \n",
       "feature                                       \n",
       "time                18         20         22  \n",
       "replicate            D          D          D  \n",
       "Gene_Symbol                                   \n",
       "Gnai3         5.402904   5.119890   4.873039  \n",
       "Cdc45        -0.308212  -1.218384  -0.541827  \n",
       "Apoh         10.468625  10.704931  10.591210  \n",
       "Narf          2.390108   1.941836   2.128767  \n",
       "Cav2          0.624375   0.806800   0.524357  \n",
       "\n",
       "[5 rows x 144 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat = pd.read_csv(\"./GSE73554_WT_AL_Intron_Exon_RFP.txt\",sep='\\t')\n",
    "dat.index = dat['Gene_Symbol']\n",
    "dat = dat.drop(['Gene_Symbol','Gene_Ensembl'], axis=1)\n",
    "dat.columns = dat.columns.str.split('_', expand = True)\n",
    "dat.columns.names = ['condition','feeding','feature','time','replicate']\n",
    "\n",
    "dat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18464198",
   "metadata": {},
   "source": [
    "#### Assess rhythmicity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "86c02887",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to infer the p-value, phase, amplitude and mean from a time-serie y with size N, period T and sampling Ts. \n",
    "def return_amp_phase_pv(y, Ts, T, N):\n",
    "    \n",
    "    # we do the harmonic regression using a Fourier series\n",
    "    t = np.linspace(0.0, N * Ts, N)\n",
    "    x_fft = np.fft.fft(y)\n",
    "    freq = np.fft.fftfreq(len(y), d=Ts)\n",
    "    index, = np.where(np.isclose(freq, 1/T, atol=0.005))\n",
    "    amp = 4 / N * np.abs(x_fft[index[0]])\n",
    "    phase = T * np.arctan2(-x_fft[index[0]].imag, x_fft[index[0]].real) / (2 * np.pi)\n",
    "    mu = 1 / N * x_fft[0].real\n",
    "    \n",
    "    #compute the residuals and statistics of the fit (pval)\n",
    "    x_hat = mu + 0.5 * amp * np.cos(2 * np.pi / T * t - 2 * np.pi * phase / T)\n",
    "    res = y - x_hat\n",
    "    sig2_1 = np.var(res)\n",
    "    sig2 = np.var(y)\n",
    "    R2 = 1 - sig2_1 / sig2\n",
    "    p = 3\n",
    "    pval = 1 - beta.cdf(R2, (p - 1) / 2, (N - p) / 2)\n",
    "    if phase < 0:\n",
    "        phase += T\n",
    "        \n",
    "    return amp, phase, pval, mu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9ea44eb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Apply the function for all the genes at the intron or exon level and retrieve the amplitude, phase, mean and p-value.\n",
    "\n",
    "Ts = 2.0 # sampling time \n",
    "T = 24 # period\n",
    "N = 48 # number of samples\n",
    "\n",
    "amp_intron = []\n",
    "phase_intron = [] \n",
    "pv_intron = []\n",
    "mu_intron = []\n",
    "\n",
    "i_pos = dat.columns.get_level_values('feature').isin(['Intron'])\n",
    "\n",
    "for i, row in dat.iterrows():\n",
    "    [a, p, pv, mu] = return_amp_phase_pv(np.array(row)[i_pos], Ts, T, N)\n",
    "    amp_intron.append(a)\n",
    "    phase_intron.append(p)\n",
    "    pv_intron.append(pv)\n",
    "    mu_intron.append(mu)\n",
    "    \n",
    "amp_exon = []\n",
    "phase_exon = [] \n",
    "pv_exon = []\n",
    "mu_exon = []\n",
    "e_pos = dat.columns.get_level_values('feature').isin(['Exon'])\n",
    "\n",
    "for i, row in dat.iterrows():\n",
    "    [a, p, pv, mu]=return_amp_phase_pv(np.array(row)[e_pos], Ts, T, N)\n",
    "    amp_exon.append(a)\n",
    "    phase_exon.append(p)\n",
    "    pv_exon.append(pv)\n",
    "    mu_exon.append(mu)\n",
    "\n",
    "#convert the lists to numpy arrays\n",
    "phase_intron=np.array(phase_intron)\n",
    "amp_intron=np.array(amp_intron)\n",
    "pv_intron=np.array(pv_intron)\n",
    "mu_intron=np.array(mu_intron)\n",
    "\n",
    "phase_exon=np.array(phase_exon)\n",
    "amp_exon=np.array(amp_exon)\n",
    "pv_exon=np.array(pv_exon)\n",
    "mu_exon=np.array(mu_exon)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de1c6f26",
   "metadata": {},
   "source": [
    "### Question 2\n",
    "1. Study the phase relationship between the pre-mRNA and mRNA for genes that you selected to be rhythmic at both the pre-mRNA and mRNA levels. Use cutoffs that give you several hundred genes.\n",
    "2. Show representative profiles of genes with short or large delays.\n",
    "3. Analyze the phase difference in function of the ratio of relative amplitudes of the mRNA and pre-mRNA. What is the expected behavior for genes with constant degradation (see course)?  \n",
    "4. For genes with a positive difference bewteen 0h and 6h, describe whether these follow the prediction.\n",
    "5. Identify few outliers and discuss them. Can you say something about how these might be regulated?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fed424ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Note that because you fitted the log2 data, the relative amplitudes take the form\n",
    "relamp_exon = 2**amp_exon[ii] - 2**(-amp_exon[ii])\n",
    "relamp_intron = 2**amp_intron[ii]- 2**(-amp_intron[ii])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
